onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -839,58 +839,102 @@ common::Status ConstantNodeProtoToTensorProto(const ONNX_NAMESPACE::NodeProto& n
   return Status::OK();
 }
 
-template <typename T>
 static Status CopySparseData(size_t n_sparse_elements,
                              const ONNX_NAMESPACE::TensorProto& indices,
                              gsl::span<const int64_t> dims,
                              std::function<void(size_t from_idx, size_t to_idx)> copier) {
   Status status = Status::OK();
   TensorShape indices_shape(indices.dims().data(), indices.dims().size());
+  const auto elements = gsl::narrow<size_t>(indices_shape.Size());
 
-  ORT_RETURN_IF_NOT(indices.data_type() == ONNX_NAMESPACE ::TensorProto_DataType_INT64, "Indicies expected to be INT64");
-
+  std::vector<int64_t> indices_values;  // used for conversion of smaller size indices
   gsl::span<const int64_t> indices_data;
-  const auto elements = static_cast<size_t>(indices_shape.Size());
-  if (indices.int64_data_size() > 0) {
-    indices_data = gsl::make_span<const int64_t>(indices.int64_data().data(), elements);
-  } else if (indices.has_raw_data()) {
-    ORT_RETURN_IF_NOT(indices.raw_data().size() == (elements * sizeof(int64_t)),
-                      "Sparse Indicies raw data size does not match expected.");
-    indices_data = gsl::make_span<const int64_t>(reinterpret_cast<const int64_t*>(indices.raw_data().data()), elements);
-  } else {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_GRAPH, "Invalid SparseTensor indices. Should either have raw or int64 data");
+  const bool has_raw_data = indices.has_raw_data();
+  switch (indices.data_type()) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64:
+      if (has_raw_data) {
+        ORT_RETURN_IF_NOT(indices.raw_data().size() == (elements * sizeof(int64_t)),
+                          "Sparse Indices raw data size does not match expected.");
+        indices_data = gsl::make_span(reinterpret_cast<const int64_t*>(indices.raw_data().data()), elements);
```

Comment:
This may cause alignment issue, you will need to unpack the data instead of directly using the tensor's raw_data